File: code/backend-python/app/core/ports/auth/guest_persistence_port.py
from abc import ABC, abstractmethod
from typing import Optional, Dict, Any

class GuestPersistencePort(ABC):
    @abstractmethod
    def create_guest_user(self, user_uuid: str, username: str, guest_cookie_token: str, expires_at_iso: str) -> int:
        pass

    @abstractmethod
    def find_guest_by_cookie_token(self, guest_cookie_token: str) -> Optional[Dict[str, Any]]:
        pass

    @abstractmethod
    def store_refresh_token(self, user_id: int, refresh_token: str, expires_at: str) -> None:
        pass

    @abstractmethod
    def update_last_access(self, user_id: int) -> None:
        pass

    @abstractmethod
    def delete_expired_guests(self) -> int:
        pass

File: code/backend/java/core/src/main/java/games/paths/core/port/auth/GuestPersistencePort.java
package games.paths.core.port.auth;

import java.util.Map;

/**
 * GuestPersistencePort - Outbound port for guest user persistence operations.
 * Implemented by a database adapter (SQLite/PostgreSQL) to store and retrieve guest data.
 */
public interface GuestPersistencePort {

    /**
     * Creates a new guest user in the database.
     *
     * @param uuid             the unique UUID for the guest user
     * @param username         the generated guest username (e.g., "guest_abc12345")
     * @param guestCookieToken the cookie token for session resumption
     * @param guestExpiresAt   the ISO-8601 timestamp when the guest session expires
     * @return the database-generated user ID
     */
    long createGuestUser(String uuid, String username, String guestCookieToken, String guestExpiresAt);

    /**
     * Finds a guest user by their cookie token.
     *
     * @param guestCookieToken the cookie token to search for
     * @return a map with user fields (id, uuid, username, guest_expires_at), or null if not found
     */
    Map<String, Object> findGuestByCookieToken(String guestCookieToken);

    /**
     * Stores a refresh token for a user.
     *
     * @param userId       the user ID
     * @param refreshToken the refresh token string
     * @param expiresAt    the ISO-8601 timestamp when the token expires
     */
    void storeRefreshToken(long userId, String refreshToken, String expiresAt);

    /**
     * Updates the last access timestamp for a user.
     *
     * @param userId the user ID
     */
    void updateLastAccess(long userId);

    /**
     * Deletes expired guest users and their associated tokens.
     *
     * @return the number of guest users removed
     */
    int deleteExpiredGuests();
}
